File: src/WN_G.cpp
#include <iostream>
#include <cstring>
#include <fstream>
#include <string>
#include <climits>
#include <algorithm>
#include <regex>

using namespace std;

string* idiomas = new string[5] { "cat", "eus", "glg","por","spa"};

string do_replace( string const & in, string const & from, string const & to )
{
  return regex_replace( in, regex(from), to );
}

int codificacionCategorias(string c)
{
    if(c == "n") return 1;
    else if (c == "v") return 2;
    else if (c == "a") return 3;
    else if (c == "r") return 4;
    else return -1;
}

int main()
{   
    for(int idiomaIndex = 0; idiomaIndex < 5; idiomaIndex++)
    {
        string datos [10];
        fstream ficheroLectura;
        fstream ficheroEscritura;
        size_t i = 0;
        size_t j = 0;

        cout << "Abriendo ficheros...\n";

        ficheroLectura.open("../mcr/"+idiomas[idiomaIndex]+"WN/wei_"+idiomas[idiomaIndex]+"-30_synset.tsv", ios::in);    
        ficheroEscritura.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_g.pl", ios::out);

        if(!ficheroLectura.is_open() || !ficheroEscritura.is_open())
        {
            cout << "No se han abierto correctamente\n";
            return 0;
        }

        cout << "Se han abierto correctamente.\n" << "Parseando a Prolog...\n";

        string line;

        while(getline(ficheroLectura, line))
        {
            i = 0;
            char *cadena = line.data();
            char *token =strtok(cadena, "\t");

            while(token != NULL)
            {
                datos[i] = token;
                token = strtok(NULL, "\t");
                i++;
            }
            ficheroEscritura << "g(" << codificacionCategorias(datos[0].substr(16,1)) << datos[0].substr(7, 8) << ",'" << do_replace(datos[6], "'", "\"") << "')." << endl;
        }

        cout << "Idioma "+idiomas[idiomaIndex]+" parseado correctamente.\n";
        ficheroLectura.close();
        ficheroEscritura.close();
    }

    cout << "¡Proceso finalizado!";

    return 0;
}

File: src/WN_G.py
import pandas as pd
import time

idiomas = ['cat', 'eng', 'eus', 'glg', 'por', 'spa']

def codificacionCategoria(c):
    categorias = {
        'n':'1',
        'v':'2',
        'a':'3',
        'r':'4'
        }
    return categorias[c]

def reemplazarComillas(palabra):
    if type(palabra) == str:
        return palabra.replace('\'','\'\'')

def crearSynsetID(synset):
    synset = codificacionCategoria(synset[-1])+synset[7:15]
    return int(synset)

for i in idiomas:
    inicio = time.time()
   
    ruta = "mcrCSV//"+i+"WN\wei_"+i+"-30_synset.csv"
    print("Leyendo dataframe "+ ruta)
    
    df = pd.read_csv(ruta, index_col=[0])
    df = df.drop(columns=['PoS', 'Desc', 'MaxNiv', 'Niv', 'Mark'])
    df.columns = ['Synset','Glosa']

    df['Synset'] = df['Synset'].apply(crearSynsetID)
    df['Glosa'] = df['Glosa'].apply(reemplazarComillas)
    
    df = df.sort_values(by=['Synset'])
    df = df.reset_index(drop = True)
    df.to_csv(i+"\PrologCSV\wn_g.csv")

    ficheroEscritura = open(i+"\Prolog\wn_g.pl", "w", encoding='utf-8')

    for index in df.index:
        ficheroEscritura.write("g("+str(df["Synset"][index])+",\'"+str(df["Glosa"][index])+"\').\n")
    
    final = time.time()

    ficheroEscritura.close()

    print('Proceso en ruta: ' + ruta + ' finalizado. Ha tardado '+str(final-inicio)+'.\n')